### PR TITLE
Implement environment call properly

### PR DESCRIPTION
**Pull request by Risheng1128**
Implement ecall instruction with control and status registers (CSRs) 
to pass the riscv-arch-test with all privilege tests.

riscv-arch-test result:
```shell
Check ebreak                    ... OK 
Check ecall                     ... OK 
Check misalign1-jalr-01         ... OK 
Check misalign2-jalr-01         ... OK 
Check misalign-beq-01           ... OK 
Check misalign-bge-01           ... OK 
Check misalign-bgeu-01          ... OK 
Check misalign-blt-01           ... OK 
Check misalign-bltu-01          ... OK 
Check misalign-bne-01           ... OK 
Check misalign-jal-01           ... OK 
Check misalign-lh-01            ... OK 
Check misalign-lhu-01           ... OK 
Check misalign-lw-01            ... OK 
Check misalign-sh-01            ... OK 
Check misalign-sw-01            ... OK
```

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
I failed to run `make check` because of no progress.

File: src/emulate.c [hunk 1]
```diff
@@ -61,9 +72,15 @@ enum {
         const uint32_t code = GET_EXCEPTION_CODE(type);                     \
         /* mepc  (Machine Exception Program Counter)                        \
          * mtval (Machine Trap Value Register)                              \
+         * mcause (Machine Cause Register): store exception code            \
          */                                                                 \
         rv->csr_mepc = rv->PC;                                              \
         rv->csr_mtval = mtval;                                              \
+        rv->csr_mcause = code;                                              \
+        if (!rv->csr_mtvec) { /* elf doesn't implement csr instruction */   \
```
Comment (jserv) [inline src/emulate.c]: The comment was misleading. You can say,
> in case CSR is not configured.

File: src/emulate.c [hunk 2]
```diff
@@ -40,14 +40,25 @@ extern struct target_ops gdbstub_ops;
     _(breakpoint)       /* Breakpoint */                     \
     _(load_misaligned)  /* Load address misaligned */        \
     _(load_fault)       /* Load access fault */              \
-    _(store_misaligned) /* Store/AMO address misaligned */
+    _(store_misaligned) /* Store/AMO address misaligned */   \
+    _(store_fault)      /* Store/AMO access fault */         \
+    _(ecall_U)          /* Environment call from U-mode */   \
+    _(ecall_S)          /* Environment call from S-mode */   \
+    _(reserved)         /* Reserved */                       \
```
Comment (jserv) [inline src/emulate.c]: Shall we care about "Reserved" item?

Comment (Risheng1128) [reply src/emulate.c]: I think the "reserved" item should be removed. Maybe I need to
modify the exception list with assigning exception code and remove
unuseful items.

Comment (jserv) [reply src/emulate.c]: In addition, check ecall_{U,S} as well since only `ecall_M` makes sense now.

**Pull request merged**